### ISSUE TITLE
GEOS-Chem 13.3.0 memory error

### ISSUE THREAD

**Issue by mvermeuel** [geoschem/geos-chem#1832]
### Name and Institution (Required)

Name: Michael Vermeuel
Institution: University of Minnesota

### Confirm you have reviewed the following documentation

- [x] [Support guidelines](https://geos-chem.readthedocs.io/en/stable/reference/SUPPORT.html)
- [X] [User manuals](https://geos-chem.readthedocs.io/en/stable/geos-chem-shared-docs/supplemental-guides/related-docs.html)
- [X] [Debugging GEOS-Chem and HEMCO errors](https://geos-chem.readthedocs.io/en/stable/geos-chem-shared-docs/supplemental-guides/debug-guide.html)
- [X] [Current and past Github issues](https://github.com/geoschem/geos-chem/issues)

### Description of your issue or question
Hello, 

I am trying to run a 4 x 5 degree, 72 level global run using MERRA2 meteorology in GEOS-Chem v13.3.0 and I am running into a segmentation error. The program ends shortly after it begins time stepping so it looks like a memory issue. I have followed the instructions to request more memory by changing the KPM_STACKSIZE in my bashrc and adding OMP_STACKSIZE to my batch file and changing num_threads, cores used, and memory requested in my batch file. I ran the same program in debug mode and it failed after reading out the restart file input so I tried a different restart file and that also failed. In the .out file for the debug I get a failure on line 2416 of fullchem_mod.F90 which is taking the square root of the molecular weight for KPP species and the error "forrtl: error (65): floating invalid" so I thought there might be a divide by zero error. I checked my .yml file to see if there were any MW errors but I don't see any so maybe that's just when the memory issue occurs. I have also tried just rebuilding the program from scratch and I still run into an issue.

Another maybe helpful note is that this is very similar to a simulation I have done before that works but the only difference is that I use MERRA2 in this run that doesn't work instead of GEOS-FP.

I have attached some documents that might be helpful: my input file, HEMCO configuration, batch file, bashrc, species database, fullchem_mod, and outputs for both the normal and debug mode. 

Thanks for any help with this.

Michael

[bashrc.txt](https://github.com/geoschem/geos-chem/files/11761505/bashrc.txt)
[GC_debug.log](https://github.com/geoschem/geos-chem/files/11761516/GC_debug.log)
[GC.log](https://github.com/geoschem/geos-chem/files/11761517/GC.log)
[HEMCO.log](https://github.com/geoschem/geos-chem/files/11761519/HEMCO.log)
[HEMCO_Config.txt](https://github.com/geoschem/geos-chem/files/11761524/HEMCO_Config.txt)
[HEMCO_debug.log](https://github.com/geoschem/geos-chem/files/11761525/HEMCO_debug.log)
[input.txt](https://github.com/geoschem/geos-chem/files/11761538/input.txt)
[run_GEOS.txt](https://github.com/geoschem/geos-chem/files/11761540/run_GEOS.txt)
[fullchem_mod.txt](https://github.com/geoschem/geos-chem/files/11761565/fullchem_mod.txt)
[slurm_debugout.txt](https://github.com/geoschem/geos-chem/files/11761551/slurm_debugout.txt)
[species_databaseyml.txt](https://github.com/geoschem/geos-chem/files/11761558/species_databaseyml.txt)
[slurm_nodebugout.txt](https://github.com/geoschem/geos-chem/files/11761580/slurm_nodebugout.txt)

**Comment by yantosca** [comment]
Thanks for your patience @mvermeuel.  Our computational cluster just underwent an OS upgrade so our attention was focused on that in these days.

If you look at the `slurm.debugout.txt` file you can see:
```console
forrtl: error (65): floating invalid
Image              PC                Routine            Line        Source             
gcclassic          00000000011C556E  Unknown               Unknown  Unknown
libpthread-2.17.s  00002B3604226630  Unknown               Unknown  Unknown
gcclassic          00000000004D898B  fullchem_mod_mp_i        2416  fullchem_mod.F90
gcclassic          000000000044B367  chemistry_mod_mp_        1314  chemistry_mod.F90
gcclassic          0000000000407FE4  MAIN__                    795  main.F90
gcclassic          0000000000405A8E  Unknown               Unknown  Unknown
libc-2.17.so       00002B3604455555  __libc_start_main     Unknown  Unknown
gcclassic          00000000004059A9  Unknown               Unknown  Unknown
real 176.85
user 1274.54
sys 51.91
srun: error: cn1082: task 0: Exited with exit code 6
```
The `floating invalid` error indicates that a number has gone out of range at line 2416 of fullchem_mod.F90.  As stated in our [supplemental guide on understanding error messages](https://geos-chem.readthedocs.io/en/latest/geos-chem-shared-docs/supplemental-guides/error-guide.html#floating-invalid-or-floating-point-exception-error) this could be caused by several things (e.g. division by zero, etc).    

If you have the gdb debugger, and you have a core file output, you can type at the command line:
```console
$ gdb gcclassic core.12345
```
(`12345` is the suffix of the core file, which is actually the Linux process ID.  Your number will be different).
This will take you directly to the point of failure.  You can then type `where` and also print out values with e.g. `print x`, where `x` is any variable name.

You might also try compiling with gfortran (any version higher than 6 will do), as in our experience, gfortran has better error-checking capability than ifort.

**Comment by mvermeuel** [comment]
Thanks @yantosca. The issue was that I was not defining a dummy species to track the loss rate of H2O2 (LH2O2) in my species_database.yml file but I was in my restart file. The program gave LH2O2 a molecular weight of -9999 which then gave an error when trying to take the square root. 

I'm not sure why this has happened in this instance since I have used the same GC version, restart file, and .yml file for other simulations that have ran successfully and showed no issue with H2O2 or LH2O2. The only difference was that I changed the meteorology from GEOS-FP to MERRA2. 

Either way, the problem is solved and this was helpful for future debugging. Thanks for helping with this!

Michael

**Comment by yantosca** [comment]
Thanks @mvermeuel.  Maybe what we can do going forward is to put a check at startup to make sure that none of the molecular weights are negative.  That would be an easy error trap in I think `GeosCore/species_database_mod.F90`.

**Comment by yantosca** [comment]
An error check for missing molecular weights is now included in PR #1843.

**Comment by yantosca** [comment]
PR #1843 is now merged into the GEOS-Chem 14.2.1 development stream.  We can now close this issue.
